src/ipc/StoreMap.h

```diff
@@ -248,11 +248,19 @@ class StoreMap
     const Anchor &peekAtEntry(const sfileno fileno) const;
 
     /// free the entry if possible or mark it as waiting to be freed if not
-    void freeEntry(const sfileno fileno);
+    /// \param stateChanged when provided, true if the entry state changes, false otherwise
+    void freeEntry(const sfileno fileno, bool *stateChanged = nullptr);
```

Comment:
If my understanding is correct, the current freeEntry() should be fixed to take into account entry 'emptiness'.Also, there should be no sense in setting waitingToBeFreed for already empty entries:

```
 bool isEmpty = s.empty();
 if (s.lock.lockExclusive()) {
        freeChain(fileno, s, false); // whether freeChain() should be called for empty entries at all?
        return !isEmpty;
 } else if (!isEmpty) {
        auto expected = false;
        // mark to free it later, if not marked yet
        return s.waitingToBeFreed.compare_exchange_strong(expected, true);
 }
  return false;
```

> whether the entry was neither empty nor marked

Then this (updated) method comment looks a bit strange and unclear. For example, if we have non-empty entry, its 'state' is changed within first 'if' block and we should return true. We do not care about 'marking' in this case.

Reply:
> the current freeEntry() should be fixed to take into account entry 'emptiness'.

To avoid misunderstanding, I am not aware of any freeEntry() bugs in the official code. AFAICT, it already handles empty entries correctly. Its reviewed PR version had minor problems with avoidable excessive `true` results, and I offered a sketch that may fix those. The sketch you offered in response is conceptually wrong as discussed immediately below.

> Also, there should be no sense in setting waitingToBeFreed for already empty entries

That may be true (I would need to research this more to provide a definitive answer), but you cannot tell whether the entry is empty unless you have an exclusive lock on it. Your sketch forgets about locking, using unreliable `isEmpty` information. You may argue that all callers should have a read lock on the entry, but I have not verified that, doubt that it is a good idea to enforce that, and even if all of that is true, then trying to lock the entry exclusively does not make sense at all, so the whole method may need to be rewritten. 

> whether freeChain() should be called for empty entries at all?

It should be or, to be more precise, this should not be changed in this PR because empty() does not necessarily guarantee absence of slots (or other entry metadata). For example, when something goes wrong during the entry formation, the writer probably has the right to simply mark the entry for deletion and release the lock. I do not know whether the current code ever does that before setting the entry key, but I think it might be a viable strategy in some cases.

Reply:
> should we mark empty locked entrires with waitingToBeFreed

For this question, I added a TODO there.

Reply:
> [is it] possible that an entry is empty but still locked?

Yes, of course. All entries must be (exclusively) locked _before_ they transition from an empty to non-empty state -- the StoreMapAnchor::key is not atomic. After locking and before setting the key, the entry is empty and locked. That time period is usually "short" but it does exist and may be "long" in rare cases. The answer seems obvious to me so perhaps I misunderstood the question?

> should freeEntry() mark empty locked entrires with waitingToBeFreed?

Yes, unless freeEntry() can actually "free" the entry (i.e., revert the entry back to its default/initial state), which is what freeChain() does. freeEntry() must either free the entry or mark it as waiting to be freed. That requirement is unrelated to whether the entry was locked and/or empty when freeEntry() was called.

Reply:
> The answer seems obvious to me so perhaps I misunderstood the question

Probably. I did not think about initialization phase(when the entry is empty some time) but about cleanup phase, when entry is _already_ empty but _still_ locked. We should not set waitingToBeFreed in this case: the entry is already freed. For example, StoreMap::abortWriting() calls freeChain() for exclusively locked entries. If later somebody else calls freeEntry(), then it changes waitingToBeFreed to 'true', returning 'true' to the caller.

Reply:
> We should not set waitingToBeFreed [when entry is already empty but still exclusively locked by cleanup code]: the entry is already freed.

AFAIK, it is impossible to reliably detect free entries exclusively locked by others, so we have to pick between not marking some entries that should be marked (a bug) and marking some entries that do not have to be marked (a performance penalty). We naturally pick the latter.

> StoreMap::abortWriting() calls freeChain() for exclusively locked entries. If later somebody else calls freeEntry(), then it changes waitingToBeFreed to 'true', returning 'true' to the caller.

* Yes, assuming that by "later" you meant "while abortWriting() is still holding the exclusive lock". See above for why we cannot prevent this excessive marking and notifications. We could probably reduce thise excess by adding more (atomic) flags, but I doubt the increased complexity is worth the performance savings because such locking conflicts should be very rare.
* The same reasoning applies to cases where we failed to exclusively lock the entry because an exclusive lock is hold by others than abortWriting().
* The same reasoning applies to cases where we failed to exclusively lock the entry because a non-exclusive lock is hold by others. In this scenario, we could reduce excessive notifications by locking the entry for _reading_ and returning false if the locked entry is empty. Please feel free to add that optimization if you think it is worth it. Please note that we should still set waitingToBeFree in this case because "empty" does not mean "free" -- we still want the next entry owner to call freeChain(), to get the entry into the "free" state.
* Finally, if the entry is not locked anymore, then the "later" code would be able to lock the entry and discover that it is already empty, returning false. No excessive notifications will happen in this case.

The only case the above list does not cover (AFAIK) is when the caller already holds a lock on the entry and, hence, does not need to lock it to test for emptiness. We could reduce excessive marking/notifications in that case but it would require splitting freeEntry() into two different methods and somehow making sure that the caller picks the right method depending on whether the caller holds a lock. Since I believe that excessive notifications ought be very rare, I do not think their reduction is worth the increased API complexity.

Please let me know if I missed any important cases or good optimization opportunities.